File: hss_okta_direct_auth/ios/Classes/HssOktaDirectAuthPlugin.swift
import Flutter
import Foundation
import OktaOAuth2
import AuthFoundation
import OktaDirectAuth

enum HssOktaError: Error {
case configError(String)
case credentialError(String)
}


open class HssOktaDirectAuthPlugin :NSObject, FlutterPlugin,HssOktaDirectAuthPluginApi
{
    
    
    var flow : DirectAuthenticationFlow?
    var status : DirectAuthenticationFlow.Status?
    
    
    func mfaOtpSignInWithCredentials(otp: String, completion: @escaping (Result<HssOktaDirectAuthResult?, Error>) -> Void) {
        Task{
            do{
             status = try await flow?.resume(self.status!, with: .otp(code: otp))
                if case let .success(token) = status{
                    Credential.default = try Credential.store(token)
                    completion(.success(HssOktaDirectAuthResult(
                        result: DirectAuthResult.success, id: token.id, token: token.idToken?.rawValue ?? "", issuedAt: Int64(((token.issuedAt?.timeIntervalSince1970 ?? 0) * 1000.0).rounded()), tokenType: token.tokenType, accessToken: token.accessToken, scope: token.scope ?? "", refreshToken: token.refreshToken ?? "")))
                }else{
                    completion(.success(HssOktaDirectAuthResult(result: DirectAuthResult.error,error: "MFA Failed")))
                }
            }catch let error{
                completion(.success(HssOktaDirectAuthResult(result: DirectAuthResult.error,error: "\(error)")))
            }
        }
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger : FlutterBinaryMessenger = registrar.messenger()
        let api : HssOktaDirectAuthPluginApi & NSObjectProtocol = HssOktaDirectAuthPlugin.init()
        HssOktaDirectAuthPluginApiSetup.setUp(binaryMessenger: messenger, api: api)
    }
    
    
    func signInWithCredentials(request: HssOktaDirectAuthRequest, completion: @escaping (Result<HssOktaDirectAuthResult?, Error>) -> Void)  {
        
        if let config = try? OAuth2Client.PropertyListConfiguration(){
            flow = DirectAuthenticationFlow(issuer: config.issuer, clientId: config.clientId, scopes: config.scopes)
            
            Task{
                do{
                    
                    let status = try await flow!.start(request.username, with: .password(request.password))
                    self.status = status
                    
                    switch status{
                    case .success(let token):
                        Credential.default = try Credential.store(token)
                        
                        completion(.success(HssOktaDirectAuthResult(
                            result: DirectAuthResult.success, id: token.id, token: token.idToken?.rawValue ?? "", issuedAt: Int64(((token.issuedAt?.timeIntervalSince1970 ?? 0) * 1000.0).rounded()), tokenType: token.tokenType, accessToken: token.accessToken, scope: token.scope ?? "", refreshToken: token.refreshToken ?? ""))) 
                        
                    case .mfaRequired(_):
                        completion(.success(HssOktaDirectAuthResult(result: DirectAuthResult.mfaRequired,error: "MFA Required")))
                    }
                    
                }catch let error{
                    debugPrint(error)
                    completion(.failure(error))
                }
            }    
        }else{
            completion(.failure(HssOktaError.configError("Configuration Error, Check okta.plist")))
        }
    }
    
    func refreshDefaultToken(completion: @escaping (Result<Bool?, Error>) -> Void){
        if let credential = Credential.default{
            Task{
                do{
                    try await credential.refresh()
                }catch let error{
                    debugPrint(error)
                    completion(.failure(error))
                }
            }
            completion(.success(true))
        }else{
            completion(.success(false))
        }
        
        
    }
    
    func revokeDefaultToken(completion: @escaping (Result<Bool?, Error>) -> Void){
        if let credential = Credential.default{
            Task{
                do{
                    try await credential.revoke()
                }catch let error{
                    debugPrint(error)
                    completion(.success(false))
                }
            }
            completion(.success(true))
        }else{
            completion(.success(false))
        }
    }
    
    func getCredential(completion: @escaping (Result<HssOktaDirectAuthResult?, Error>) -> Void){
        if let result = Credential.default{
            
            var resultEnum = DirectAuthResult.success
            completion(.success(HssOktaDirectAuthResult(
                result: resultEnum, id: result.token.id, token: result.token.idToken?.rawValue ?? "", issuedAt: Int64(((result.token.issuedAt?.timeIntervalSince1970 ?? 0) * 1000.0).rounded()), tokenType: result.token.tokenType, accessToken: result.token.accessToken, scope: result.token.scope ?? "", refreshToken: result.token.refreshToken ?? "")))
        }else{
            var resultEnum = DirectAuthResult.error
            completion(.success(HssOktaDirectAuthResult(result: resultEnum,error: "Failed to Login : Server did not provide result")))
        }
    }
    
   
    
    
  
    
}

File: hss_okta_direct_auth/android/src/main/kotlin/dev/hypersense/software/hss_okta_direct_auth/HssOktaDirectAuthPlugin.kt
package dev.hypersense.software.hss_okta_direct_auth


import android.content.Context
import com.okta.authfoundation.claims.email
import com.okta.authfoundation.claims.familyName
import com.okta.authfoundation.claims.gender
import com.okta.authfoundation.claims.givenName
import com.okta.authfoundation.claims.issuedAt
import com.okta.authfoundation.claims.middleName
import com.okta.authfoundation.claims.phoneNumber
import com.okta.authfoundation.claims.scope
import com.okta.authfoundation.claims.userId
import com.okta.authfoundation.claims.username
import com.okta.authfoundation.client.OidcClient
import com.okta.authfoundation.client.OidcClientResult
import com.okta.authfoundation.client.OidcConfiguration
import com.okta.authfoundation.credential.CredentialDataSource.Companion.createCredentialDataSource
import com.okta.authfoundationbootstrap.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
//import com.okta.idx.kotlin.*
import com.okta.oauth2.ResourceOwnerFlow.Companion.createResourceOwnerFlow
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import okhttp3.HttpUrl.Companion.toHttpUrl
import java.util.Date

class HssOktaDirectAuthPlugin : HssOktaDirectAuthPluginApi, FlutterPlugin{
    var context : Context? = null



    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        HssOktaDirectAuthPluginApi.setUp(binding.binaryMessenger, this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        HssOktaDirectAuthPluginApi.setUp(binding.binaryMessenger, null)
    }

    override fun initializeConfiguration(
        clientid: String,
        signInRedirectUrl: String,
        signOutRedirectUrl: String,
        issuer: String,
        scopes: String
    ) {

        println("Initializing OIDC Configuration")
        val oidcConfiguration = OidcConfiguration(
            clientId =  clientid,
            defaultScope = scopes,
        )

        println("${issuer.toHttpUrl()}")

       val oidcClient = OidcClient.createFromDiscoveryUrl(
            oidcConfiguration!!,
           "${issuer}/.well-known/openid-configuration".toHttpUrl(),)


        if(context == null){
            throw  Exception("Context is null")
        }

        CredentialBootstrap.initialize(oidcClient.createCredentialDataSource(context!!))

        println("Done Initializing OIDC Configuration /()")
    }


    private suspend fun signInWithCredentialsRoutine(request : HssOktaDirectAuthRequest): HssOktaDirectAuthResult {
        val flow = CredentialBootstrap.oidcClient.createResourceOwnerFlow()

        when(val res = flow.start(request.username,request.password)){
            is OidcClientResult.Error -> {
                return HssOktaDirectAuthResult(
                    result = DirectAuthResult.ERROR,
                    error = res.exception.message
                )
            }
            is OidcClientResult.Success -> {

                CredentialBootstrap.defaultCredential().storeToken(token = res.result)
                var userInfo = CredentialBootstrap.defaultCredential().getUserInfo()
                var userInfoResult = userInfo.getOrThrow()


                return HssOktaDirectAuthResult(
                    result = DirectAuthResult.SUCCESS,
                    id = res.result.idToken,
                    token = res.result.accessToken,
                    issuedAt = userInfoResult.issuedAt?.toLong(),
                    tokenType = res.result.tokenType,
                    scope = res.result.scope,
                    refreshToken = res.result.refreshToken,
                    userInfo = UserInfo(
                        userId = userInfoResult.userId?:"",
                        givenName = userInfoResult.givenName?:"",
                        middleName = userInfoResult.middleName?:"",
                        familyName = userInfoResult.familyName?:"",
                        gender = userInfoResult.gender?:"",
                        email = userInfoResult.email?:"",
                        phoneNumber = userInfoResult.phoneNumber?:"",
                        username = userInfoResult.username?:""
                    )
                )
            }
        }
    }

    override fun signInWithCredentials(
        request: HssOktaDirectAuthRequest,
        callback: (Result<HssOktaDirectAuthResult?>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
           var result = signInWithCredentialsRoutine(request)
            callback.invoke(Result.success(result))
        }
    }

    override fun mfaOtpSignInWithCredentials(
        otp: String,
        callback: (Result<HssOktaDirectAuthResult?>) -> Unit
    ) {
        TODO("Not yet implemented")
    }

    override fun refreshDefaultToken(callback: (Result<Boolean?>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            when(val result = CredentialBootstrap.defaultCredential().refreshToken()){
                is OidcClientResult.Error -> callback.invoke(Result.failure(Exception("Failed to refresh token")))
                is OidcClientResult.Success -> callback.invoke(Result.success(true))
            }
        }
    }

    override fun revokeDefaultToken(callback: (Result<Boolean?>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            when(val result = CredentialBootstrap.defaultCredential().refreshToken()){
                is OidcClientResult.Error -> callback.invoke(Result.failure(Exception("Failed to revoke token")))
                is OidcClientResult.Success -> callback.invoke(Result.success(true))
            }
        }
    }

    override fun getCredential(callback: (Result<HssOktaDirectAuthResult?>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            var credential = CredentialBootstrap.defaultCredential()
            var userInfo = credential.getUserInfo().getOrThrow()


            callback.invoke(Result.success(HssOktaDirectAuthResult(
                result = DirectAuthResult.SUCCESS,
                id = userInfo.userId,
                token = credential.token?.accessToken,
                issuedAt = userInfo.issuedAt?.toLong(),
                tokenType = credential.token?.tokenType,
                scope = credential.token?.scope,
                refreshToken = credential.token?.refreshToken,
                userInfo = UserInfo(
                    userId = userInfo.userId?:"",
                    givenName = userInfo.givenName?:"",
                    middleName = userInfo.middleName?:"",
                    familyName = userInfo.familyName?:"",
                    gender = userInfo.gender?:"",
                    email = userInfo.email?:"",
                    phoneNumber = userInfo.phoneNumber?:"",
                    username = userInfo.username?:""
                )
            )))

    }

}}